File: Blocktrader.Service/Files/Timestamp.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using Nexus.Blocktrader.Domain;

namespace Nexus.Blocktrader.Service.Files
{
    public class Timestamp
    {
        public DateTime Date { get; }

        public TickerInfo TickerInfo { get; }
        
        public Timestamp(DateTime date, [NotNull] TickerInfo tickerInfo)
        {
            Date = date;
            TickerInfo = tickerInfo;
        }

        public byte[] ToBytes()
        {
            return BitConverter.GetBytes(Date.ToBinary())
                .Concat(BitConverter.GetBytes(TickerInfo.AveragePrice))
                .Concat(BitConverter.GetBytes(TickerInfo.OrderBook.Bids.Length))
                .Concat(TickerInfo.OrderBook.Bids.SelectMany(b => b.ToBytes()))
                .Concat(BitConverter.GetBytes(TickerInfo.OrderBook.Asks.Length))
                .Concat(TickerInfo.OrderBook.Asks.SelectMany(a => a.ToBytes()))
                .ToArray();
        }
        
        public byte[] ToBytes2()
        {
            var unixTimestamp = (long)(Date.Subtract(new DateTime(1970, 1, 1))).TotalSeconds;
            unixTimestamp = unixTimestamp * 1000;
            
            return BitConverter.GetBytes(unixTimestamp)
                .Concat(BitConverter.GetBytes(TickerInfo.AveragePrice))
                .Concat(BitConverter.GetBytes(TickerInfo.OrderBook.Bids.Length))
                .Concat(TickerInfo.OrderBook.Bids.SelectMany(b => b.ToBytes()))
                .Concat(BitConverter.GetBytes(TickerInfo.OrderBook.Asks.Length))
                .Concat(TickerInfo.OrderBook.Asks.SelectMany(a => a.ToBytes()))
                .ToArray();
        }

        public static IEnumerable<Timestamp> FromBytes(byte[] bytes)
        {
            var index = 0;
            while (index < bytes.Length)
            {
                var dateTime = DateTime.FromBinary(BitConverter.ToInt64(bytes, index));
                index += 8;
                var averagePrice = BitConverter.ToSingle(bytes, index);
                index += 4;
                var bidsCount = BitConverter.ToInt32(bytes, index);
                index += 4;
                var bids = new List<Order>(bidsCount);
                for (var i = 0; i < bidsCount; i++)
                {
                    bids.Add(Order.FromBytes(bytes, index));
                    index += 8;
                }

                var asksCount = BitConverter.ToInt32(bytes, index);
                index += 4;
                var asks = new List<Order>(bidsCount);
                for (var i = 0; i < asksCount; i++)
                {
                    asks.Add(Order.FromBytes(bytes, index));
                    index += 8;
                }

                yield return new Timestamp(dateTime, new TickerInfo(averagePrice, new OrderBook(bids.ToArray(), asks.ToArray()), dateTime));
            }
        }
    }
}

File: Blocktrader.Api/ClientApp/src/Models/Timestamp.ts
export class Timestamp {
    public date: Date;
    public tickerInfo: TickerInfo;
    constructor(date: Date, tickerInfo: TickerInfo){
        this.date = date;
        this.tickerInfo = tickerInfo
    }
}

export class TickerInfo {
    public orderBook: OrderBook;
    public currentPrice: number;
    constructor(orderBook: OrderBook, currentPrice: number) {
        this.orderBook = orderBook;
        this.currentPrice = currentPrice;
    }
}

export class OrderBook {
    public bids: Order[];
    public asks: Order[];
    constructor(bids: Order[], asks: Order[]){
        this.bids = bids;
        this.asks = asks;
    }
}

export class Order {
    public price: number;
    public amount: number;
    constructor(price: number, amount: number){
        this.price = price;
        this.amount = amount;
    }
}

export function timestampFromBytes(buffer: ArrayBuffer) {
    const result = [];

    let index = 0;
    const dataView = new DataView(buffer);

    console.log(buffer.byteLength);
    while (index < buffer.byteLength) {
        const date = new Date(Number(dataView.getBigInt64(index, true)));
        index += 8;
        const averagePrice = dataView.getFloat32(index, true);
        index += 4;

        const bidsCount = dataView.getInt32(index, true);
        index += 4;
        const bids = [];
        for (let i = 0; i < bidsCount; i++) {
            bids.push(getOrderFromBytes(dataView, index));
            index += 8;
        }

        const asksCount = dataView.getInt32(index, true);
        index += 4;
        const asks = [];
        for (let i = 0; i < asksCount; i++) {
            asks.push(getOrderFromBytes(dataView, index));
            index += 8;
        }

        result.push(new Timestamp(date, new TickerInfo(new OrderBook(bids, asks), averagePrice)));
    }

    return result;
}

function getOrderFromBytes(dataView: DataView, index: number) {
    return new Order(
        dataView.getFloat32(index, true),
        dataView.getFloat32(index + 4, true));
}
